### ISSUE TITLE
Prebuilt binaries for Windows

### ISSUE THREAD

**Issue by SA-j00u** [artyom-beilis/pytorch_dlprim#75]
is there any prebuilt binaries for windows?
i tried to build it with mingw
but got MSVC related errors (in torch\share\cmake\Caffe2)
after it i get another errors...

i tried torch-1.13.1+cpu and torch-2.0.1+cpu

**Comment by artyom-beilis** [comment]
What are the errors? I think you need to use MSVC.

**Comment by SA-j00u** [comment]
> What are the errors? I think you need to use MSVC.

trying llvm-mingw
get same errors
![2024-05-03_165542_](https://github.com/artyom-beilis/pytorch_dlprim/assets/166461727/ec181626-e32b-41d3-9efb-3dc3c1531bfd)

**Comment by SA-j00u** [comment]
```
ld.lld: [0;31merror: [0mundefined symbol: unsigned short caffe2::TypeMeta::add
TypeMetaData<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1:
:allocator<char>>>()
>>> referenced by objects.a(pointwise_ops.cpp.obj):(unsigned short caffe2::TypeM
eta::_typeMetaData<std::__1::basic_string<char, std::__1::char_traits<char>, std
::__1::allocator<char>>>())
>>> referenced by objects.a(tensor_ops.cpp.obj)
>>> referenced by objects.a(vision_ops.cpp.obj)

ld.lld: [0;31merror: [0mundefined symbol: unsigned short caffe2::TypeMeta::add
TypeMetaData<unsigned short>()
>>> referenced by objects.a(pointwise_ops.cpp.obj):(unsigned short caffe2::TypeM
eta::_typeMetaData<unsigned short>())
>>> referenced by objects.a(tensor_ops.cpp.obj)
>>> referenced by objects.a(vision_ops.cpp.obj)

ld.lld: [0;31merror: [0mundefined symbol: unsigned short caffe2::TypeMeta::add
TypeMetaData<char>()
>>> referenced by objects.a(pointwise_ops.cpp.obj):(unsigned short caffe2::TypeM
eta::_typeMetaData<char>())
>>> referenced by objects.a(tensor_ops.cpp.obj)
>>> referenced by objects.a(vision_ops.cpp.obj)

ld.lld: [0;31merror: [0mundefined symbol: unsigned short caffe2::TypeMeta::add
TypeMetaData<std::__1::unique_ptr<std::__1::mutex, std::__1::default_delete<std:
:__1::mutex>>>()
>>> referenced by objects.a(pointwise_ops.cpp.obj):(unsigned short caffe2::TypeM
eta::_typeMetaData<std::__1::unique_ptr<std::__1::mutex, std::__1::default_delet
e<std::__1::mutex>>>())
>>> referenced by objects.a(tensor_ops.cpp.obj)
>>> referenced by objects.a(vision_ops.cpp.obj)

ld.lld: [0;31merror: [0mundefined symbol: unsigned short caffe2::TypeMeta::add
TypeMetaData<std::__1::unique_ptr<std::__1::atomic<bool>, std::__1::default_dele
te<std::__1::atomic<bool>>>>()
>>> referenced by objects.a(pointwise_ops.cpp.obj):(unsigned short caffe2::TypeM
eta::_typeMetaData<std::__1::unique_ptr<std::__1::atomic<bool>, std::__1::defaul
t_delete<std::__1::atomic<bool>>>>())
>>> referenced by objects.a(tensor_ops.cpp.obj)
>>> referenced by objects.a(vision_ops.cpp.obj)

ld.lld: [0;31merror: [0mundefined symbol: unsigned short caffe2::TypeMeta::add
TypeMetaData<std::__1::vector<int, std::__1::allocator<int>>>()
>>> referenced by objects.a(pointwise_ops.cpp.obj):(unsigned short caffe2::TypeM
eta::_typeMetaData<std::__1::vector<int, std::__1::allocator<int>>>())
>>> referenced by objects.a(tensor_ops.cpp.obj)
>>> referenced by objects.a(vision_ops.cpp.obj)

ld.lld: [0;31merror: [0mundefined symbol: unsigned short caffe2::TypeMeta::add
TypeMetaData<std::__1::vector<long long, std::__1::allocator<long long>>>()
>>> referenced by objects.a(pointwise_ops.cpp.obj):(unsigned short caffe2::TypeM
eta::_typeMetaData<std::__1::vector<long long, std::__1::allocator<long long>>>(
))
>>> referenced by objects.a(tensor_ops.cpp.obj)
>>> referenced by objects.a(vision_ops.cpp.obj)

ld.lld: [0;31merror: [0mundefined symbol: unsigned short caffe2::TypeMeta::add
TypeMetaData<std::__1::vector<unsigned long, std::__1::allocator<unsigned long>>
>()
>>> referenced by objects.a(pointwise_ops.cpp.obj):(unsigned short caffe2::TypeM
eta::_typeMetaData<std::__1::vector<unsigned long, std::__1::allocator<unsigned
long>>>())
>>> referenced by objects.a(tensor_ops.cpp.obj)
>>> referenced by objects.a(vision_ops.cpp.obj)

ld.lld: [0;31merror: [0mundefined symbol: unsigned short caffe2::TypeMeta::add
TypeMetaData<bool*>()
>>> referenced by objects.a(pointwise_ops.cpp.obj):(unsigned short caffe2::TypeM
eta::_typeMetaData<bool*>())
>>> referenced by objects.a(tensor_ops.cpp.obj)
>>> referenced by objects.a(vision_ops.cpp.obj)

ld.lld: [0;31merror: [0mundefined symbol: unsigned short caffe2::TypeMeta::add
TypeMetaData<char*>()
>>> referenced by objects.a(pointwise_ops.cpp.obj):(unsigned short caffe2::TypeM
eta::_typeMetaData<char*>())
>>> referenced by objects.a(tensor_ops.cpp.obj)
>>> referenced by objects.a(vision_ops.cpp.obj)

ld.lld: [0;31merror: [0mundefined symbol: unsigned short caffe2::TypeMeta::add
TypeMetaData<int*>()
>>> referenced by objects.a(pointwise_ops.cpp.obj):(unsigned short caffe2::TypeM
eta::_typeMetaData<int*>())
>>> referenced by objects.a(tensor_ops.cpp.obj)
>>> referenced by objects.a(vision_ops.cpp.obj)

ld.lld: [0;31merror: [0mundefined symbol: unsigned short caffe2::TypeMeta::add
TypeMetaData<long>()
>>> referenced by objects.a(pointwise_ops.cpp.obj):(unsigned short caffe2::TypeM
eta::_typeMetaData<long>())
>>> referenced by objects.a(tensor_ops.cpp.obj)
>>> referenced by objects.a(vision_ops.cpp.obj)

ld.lld: [0;31merror: [0mundefined symbol: unsigned short caffe2::TypeMeta::add
TypeMetaData<std::__1::vector<long, std::__1::allocator<long>>>()
>>> referenced by objects.a(pointwise_ops.cpp.obj):(unsigned short caffe2::TypeM
eta::_typeMetaData<std::__1::vector<long, std::__1::allocator<long>>>())
>>> referenced by objects.a(tensor_ops.cpp.obj)
>>> referenced by objects.a(vision_ops.cpp.obj)

ld.lld: [0;31merror: [0mundefined symbol: unsigned short caffe2::TypeMeta::add
TypeMetaData<float*>()
>>> referenced by objects.a(pointwise_ops.cpp.obj):(unsigned short caffe2::TypeM
eta::_typeMetaData<float*>())
>>> referenced by objects.a(tensor_ops.cpp.obj)
>>> referenced by objects.a(vision_ops.cpp.obj)

ld.lld: [0;31merror: [0mundefined symbol: unsigned short caffe2::TypeMeta::add
TypeMetaData<c10::Half*>()
>>> referenced by objects.a(pointwise_ops.cpp.obj):(unsigned short caffe2::TypeM
eta::_typeMetaData<c10::Half*>())
>>> referenced by objects.a(tensor_ops.cpp.obj)
>>> referenced by objects.a(vision_ops.cpp.obj)

ld.lld: [0;31merror: [0mundefined symbol: __declspec(dllimport) c10::detail::t
orchCheckFail(char const*, char const*, unsigned int, char const*)
>>> referenced by objects.a(pointwise_ops.cpp.obj):(ptdlprim::isCPUScalar(at::Te
nsor const&, double&))
>>> referenced by objects.a(pointwise_ops.cpp.obj):(ptdlprim::squeeze_dim(dlprim
::Shape, c10::OptionalArrayRef<long long>, bool))
>>> referenced by objects.a(pointwise_ops.cpp.obj):(ptdlprim::sum_mean_out(at::T
ensor const&, c10::OptionalArrayRef<long long>, bool, c10::optional<c10::ScalarT
ype>, at::Tensor&, bool))
>>> referenced 91 more times

ld.lld: [0;31merror: [0mundefined symbol: __declspec(dllimport) caffe2::TypeMe
ta::error_unsupported_typemeta(caffe2::TypeMeta)
>>> referenced by objects.a(pointwise_ops.cpp.obj):(caffe2::TypeMeta::toScalarTy
pe())
>>> referenced by objects.a(tensor_ops.cpp.obj)
>>> referenced by objects.a(vision_ops.cpp.obj)

ld.lld: [0;31merror: [0mundefined symbol: __declspec(dllimport) at::_ops::copy
_::call(at::Tensor&, at::Tensor const&, bool)
>>> referenced by objects.a(pointwise_ops.cpp.obj):(at::Tensor::copy_(at::Tensor
 const&, bool) const)
>>> referenced by objects.a(tensor_ops.cpp.obj)

ld.lld: [0;31merror: [0mundefined symbol: __declspec(dllimport) torch::Library
::Library(torch::Library::Kind, std::__1::basic_string<char, std::__1::char_trai
ts<char>, std::__1::allocator<char>>, c10::optional<c10::DispatchKey>, char cons
t*, unsigned int)
>>> referenced by objects.a(pointwise_ops.cpp.obj):(torch::detail::TorchLibraryI
nit::TorchLibraryInit(torch::Library::Kind, void (*)(torch::Library&), char cons
t*, c10::optional<c10::DispatchKey>, char const*, unsigned int))
>>> referenced by objects.a(tensor_ops.cpp.obj)
>>> referenced by objects.a(vision_ops.cpp.obj)

ld.lld: [0;31merror: [0mundefined symbol: __declspec(dllimport) c10::detail::t
orchInternalAssertFail(char const*, char const*, unsigned int, char const*, std:
:__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>
 const&)
>>> referenced by objects.a(pointwise_ops.cpp.obj):(c10::Device::validate())
>>> referenced by objects.a(pointwise_ops.cpp.obj):(c10::Device::validate())
>>> referenced by objects.a(pointwise_ops.cpp.obj):(c10::IValue::isIntrusivePtr(
) const)
>>> referenced 17 more times
```

**Comment by artyom-beilis** [comment]
You need visual studio 2022 - because this is what pytorch is built with. I published instructions there: https://github.com/artyom-beilis/pytorch_dlprim/blob/master/README-build.md

If you can install  python 3.12 and pythorch 2.4 on 64 bit system - it is my test environment. So I can upload built package this time.

**Comment by artyom-beilis** [comment]
Implemented in release 0.1.0
